examples/rag/rag_agents.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+"""
+This example shows how to build an agent with RAG (backup by LlamaIndex)
+"""
+
+from typing import Optional
+from llama_index.core import SimpleDirectoryReader
+from langchain_community.document_loaders import DirectoryLoader
+
+from agentscope.prompt import PromptType
+from agentscope.agents.agent import AgentBase
+from agentscope.prompt import PromptEngine
+from agentscope.message import Msg
+from agentscope.models import load_model_by_config_name
+from agentscope.rag.llama_index_rag import LlamaIndexRAG
+from agentscope.rag.langchain_rag import LangChainRAG
+
+
+class RAGAgent(AgentBase):
+    """
+    Base class for RAG agents, child classes include the
+    RAG agents built with LlamaIndex and LangChain in this file
+    """
+
+    def __init__(
+        self,
+        name: str,
+        sys_prompt: Optional[str] = None,
+        model_config_name: str = None,
+        emb_model_config_name: str = None,
+        use_memory: bool = True,
+        memory_config: Optional[dict] = None,
+        prompt_type: Optional[PromptType] = PromptType.LIST,
+        config: Optional[dict] = None,
+    ) -> None:
+        super().__init__(
+            name=name,
+            sys_prompt=sys_prompt,
+            model_config_name=model_config_name,
+            use_memory=use_memory,
+            memory_config=memory_config,
+        )
+        # init prompt engine
+        self.engine = PromptEngine(self.model, prompt_type=prompt_type)
+        self.emb_model = load_model_by_config_name(emb_model_config_name)
+
+        # init rag as None
+        # MUST USE LlamaIndexAgent OR LangChainAgent
+        self.rag = None
+        self.config = config or {}
+
+    def reply(
+        self,
+        x: dict = None,
+    ) -> dict:
+        """
+        Reply function of the RAG agent.
+        Processes the input data,
+        1) use the input data to retrieve with RAG function;
+        2) generates a prompt using the current memory and system
+        prompt;
+        3) invokes the language model to produce a response. The
+        response is then formatted and added to the dialogue memory.
+
+        Args:
+            x (`dict`, defaults to `None`):
+                A dictionary representing the user's input to the agent. This
+                input is added to the memory if provided. Defaults to
+                None.
+        Returns:
+            A dictionary representing the message generated by the agent in
+            response to the user's input.
+        """
+        retrieved_docs_to_string = ""
+        # record the input if needed
+        if x is not None:
+            self.memory.add(x)
```

Comment:
1. Currently, we don't need to determine if x is `None` or not.  The `add` function will handle `None` value. 
2. Since in this agent, we expose the argument `use_memroy` to users, here we should determine if `self.memory` is `None`.

Reply:
updated.